chapters/en/slides/module4/module4_04.md

```diff
@@ -0,0 +1,131 @@
+
+---
+
+# Handling merge conflicts: JupyterLab
+
+How to open a Jupyter notebook as a plain text file view in Jupyter. 
+
+<center>
+
+<img src='/module4/vc-merge-conflict-4.png' width="60%" alt="404 image"/>
+
+</center>
+
+Notes: The beginning of the merge conflict is preceded by `<<<<<<< HEAD` and the end of the merge conflict is marked by `>>>>>>>`. Between these markings, Git also inserts a separator (`=======`). The version of the change before the separator is your change, and the version that follows the separator was the change that existed on GitHub. In the image, you can see that in your local repository there is a line of code that calls `scale_color_manual` with three color values (`deeppink2`, `cyan4`, and `purple1`). It looks like your collaborator made an edit to that line too, except with different colors (to `blue3`, `red3`, `and black`)!
+
+---
+
+# Handling merge conflicts: JupyterLab
+
+File where a merge conflict has been resolved. 
+
+<center>
+
+<img src='/module4/vc-merge-conflict-5.png' width="60%" alt="404 image"/>
+
+</center>
+
+Notes: Once you have decided which version of the change (or what combination!) to keep, you need to use the plain text editor to remove the special marks that Git added.
+
+We can do anything we want:
+    - keep the change made in the local repository,
+    - keep the change made in the remote repository,
+    - write something new to replace both, or get rid of the change entirely.
+
+After your changes the file must be saved, added to the staging area, and then committed before you will be able to push your changes to GitHub.
+
+---
+# Handling merge conflicts: the terminal
+
+
+<center>
+
+<img src='/module4/vc-merge-conflict-t.png' width="70%" alt="404 image"/>
+
+</center>
+
+Notes: It is possible to solve the merge conflicts using the terminal. For this you have to open the editor and find the files where the conflict is.
+You will be able to visualize the merge conflict as appears in the image.
+
+
+
+---
+# Extra: Stashing local non-committed changes before pulling (terminal)
+
+
+```
+error: Your local changes to the following files would be overwritten by merge:
+        README.md
+Please commit your changes or stash them before you merge.
+Aborting
+```
+
+Notes: 
+We have learned that if there are changes on your remote repo in GitHub and you already have local committed changes, you will need to pull before you can push. If the local and remote changes are in the same lines, you will have to resolve the resulting merge conflict, otherwise git will merge automatically. But what if you have just started to make changes to a file when you realize that you forgot to pull before you started to work? The first thing to do is to try to pull, if you're lucky there are either no new changes or they are not in the same file you modified. If they are in the same file, you will get an error message like the one in the slide.
```

Comment:
```suggestion
We have learned that if there are changes in your remote repo on GitHub and you already have locally committed changes, you will need to pull before you can push. If the local and remote changes are in the same lines, you will have to resolve the resulting merge conflict, otherwise Git will merge automatically. But what if you have just started to make changes to a file when you realize that you forgot to pull before you started to work? The first thing to do is to try to pull, if you're lucky there are either no new changes or they are not in the same file you modified. If they are in the same file, you will get an error message like the one in the slide.
```